core/src/main/scala/org/apache/spark/api/python/VirtualEnvFactory.scala

```diff
@@ -0,0 +1,164 @@
+     *
+     */
+    logInfo("Start to setup virtualenv...")
+    logDebug("user.dir=" + System.getProperty("user.dir"))
+    logDebug("user.home=" + System.getProperty("user.home"))
+
+    require(virtualEnvType == "native" || virtualEnvType == "conda",
+      s"VirtualEnvType: $virtualEnvType is not supported." )
+    require(new File(virtualEnvBinPath).exists(),
+      s"VirtualEnvBinPath: $virtualEnvBinPath is not defined or doesn't exist.")
+    // Two scenarios of creating virtualenv:
+    // 1. created in yarn container. Yarn will clean it up after container is exited
+    // 2. created outside yarn container. Spark need to create temp directory and clean it after app
+    //    finish.
+    //      - driver of PySpark shell
+    //      - driver of yarn-client mode
+    if (isLauncher ||
+      (isDriver && conf.get("spark.submit.deployMode") == "client")) {
+      val virtualenvBasedir = Files.createTempDir()
+      virtualenvBasedir.deleteOnExit()
+      virtualEnvName = virtualenvBasedir.getAbsolutePath
+    } else if (isDriver && conf.get("spark.submit.deployMode") == "cluster") {
+      virtualEnvName = "virtualenv_driver"
+    } else {
+      // use the working directory of Executor
+      virtualEnvName = "virtualenv_" + conf.getAppId + "_" + VIRTUALENV_ID.getAndIncrement()
+    }
+
+    // Use the absolute path of requirement file in the following cases
+    // 1. driver of pyspark shell
+    // 2. driver of yarn-client mode
+    // otherwise just use filename as it would be downloaded to the working directory of Executor
+    val pysparkRequirements =
+      if (isLauncher ||
+        (isDriver && conf.get("spark.submit.deployMode") == "client")) {
+        conf.getOption("spark.pyspark.virtualenv.requirements")
+      } else {
+        conf.getOption("spark.pyspark.virtualenv.requirements").map(_.split("/").last)
+      }
+
+    val createEnvCommand =
+      if (virtualEnvType == "native") {
+        List(virtualEnvBinPath,
+          "-p", pythonExec,
+          "--no-site-packages", virtualEnvName)
+      } else {
+        // Two cases of conda
+        //    1. requirement file is specified. (Batch mode)
+        //    2. requirement file is not specified. (Interactive mode).
+        //       In this case `spark.pyspark.virtualenv.python_version` must be specified.
+
+        if (pysparkRequirements.isDefined) {
+          List(virtualEnvBinPath,
+            "create", "--prefix", virtualEnvName,
+            "--file", pysparkRequirements.get, "-y")
+        } else {
+          require(conf.contains("spark.pyspark.virtualenv.python_version"),
+            "spark.pyspark.virtualenv.python_version is not set when using conda " +
+              "in interactive mode")
+          val pythonVersion = conf.get("spark.pyspark.virtualenv.python_version")
+          List(virtualEnvBinPath,
+            "create", "--prefix", virtualEnvName,
+            "python=" + pythonVersion, "-y")
```

Comment:
So we only use the pythonVersion if there is no requirements file? Why?

Reply:
It is for conda. Conda's requirement file contains python while native virtualenv don't have python. That's why user have to specify python_version when using conda, otherwise conda will create a virtualenv without python.